src/Embeddings/VectorStores/Redis/RedisVectorStore.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace LLPhant\Embeddings\VectorStores\Redis;
+
+use LLPhant\Embeddings\Document;
+use LLPhant\Embeddings\DocumentUtils;
+use LLPhant\Embeddings\VectorStores\VectorStoreBase;
+use Predis;
+use Predis\Command\Argument\Search\CreateArguments;
+use Predis\Command\Argument\Search\DropArguments;
+use Predis\Command\Argument\Search\SchemaFields\TextField;
+use Predis\Command\Argument\Search\SchemaFields\VectorField;
+use Predis\Command\Argument\Search\SearchArguments;
+
+class RedisVectorStore extends VectorStoreBase
+{
+    public const LLPHANT_INDEX = 'llphant';
+
+    public Predis\Client $client;
+
+    public function __construct(Predis\Client $client)
+    {
+        $this->client = $client;
+
+        try {
+            // Drop index and delete all data in it
+            $client->ftdropindex(self::LLPHANT_INDEX, (new DropArguments())->dd());
+        } catch (Predis\Response\ServerException $e) {
+            if ($e->getMessage() !== 'Unknown Index name') {
+                throw $e;
+            }
+        }
+    }
+
+    public function addDocument(Document $document): void
+    {
+        $this->client->jsonmset([...$this->generateRedisJsonSetArguments($document)]);
+    }
+
+    /**
+     * @param  Document[]  $documents
+     */
+    public function addDocuments(array $documents): void
+    {
+        if ($documents === []) {
+            return;
+        }
+
+        $redisArgs = [];
+        foreach ($documents as $document) {
+            array_push($redisArgs, ...$this->generateRedisJsonSetArguments($document));
+        }
+
+        $this->client->jsonmset(...$redisArgs);
+    }
+
+    /**
+     * Return docs most similar to the embedding.
+     *
+     * @param  float[]  $embedding
+     * @param  array<string, string|int>|array<mixed[]>  $additionalArguments
+     * @return mixed[]
+     */
+    public function similaritySearch(array $embedding, int $k = 4, array $additionalArguments = []): array
+    {
+        $vectorDimension = count($embedding);
+        $this->createIndexIfMissing($vectorDimension, $k);
+
+        $binaryQueryVector = '';
+        foreach ($embedding as $value) {
+            $binaryQueryVector .= pack('f', $value);
+        }
+
+        $filter = array_key_exists('filters', $additionalArguments) ?
+            $additionalArguments['filters']
+            : '*';
+
+        $rawResults = $this->client->ftsearch(
+            self::LLPHANT_INDEX,
+            "($filter)=>[KNN $k @embedding \$query_vector AS distance]",
+            (new SearchArguments())
+                ->dialect(2)
+                ->params(['query_vector', $binaryQueryVector])
+                ->sortBy('distance', 'asc')
+        );
+        $formattedResults = $this->formatRedisResults($rawResults);
+
+        return $formattedResults;
+    }
+
+    private function formatRedisResults(array $rawRedisResults)
+    {
+        $formattedResults = [];
+        for ($i = 1; $i < count($rawRedisResults); $i += 2) {
```

Comment:
maybe explicit the magic arounf $i +=2 and $rawRedisResults[$i + 1][3]

Reply:
Avec un exemple c'est ok ?
![Capture d’écran 2023-10-05 à 11 16 59](https://github.com/theodo-group/LLPhant/assets/55784802/fd3d526a-de37-4172-88b1-7313b4a68214)